src/components/application_manager/src/hmi_language_handler.cc

```diff
@@ -301,26 +310,28 @@ void HMILanguageHandler::HandleWrongLanguageApp(const Apps::value_type& app) {
       commands::Command::ORIGIN_SDL);
   application_manager_.UnregisterApplication(
       app.first, mobile_apis::Result::SUCCESS, false);
-  apps_.erase(it);
-  if (0 == apps_.size()) {
-    LOG4CXX_DEBUG(logger_,
-                  "All apps processed. Unsubscribing from all events.");
-    unsubscribe_from_all_events();
-  }
+  LOG4CXX_INFO(logger_,
+               "Unregistering application with app_id "
+                   << app.first << " because of HMI language(s) mismatch.");
 }
 
 void HMILanguageHandler::CheckApplication(const Apps::value_type app) {
   LOG4CXX_AUTO_TRACE(logger_);
-  sync_primitives::AutoLock lock(apps_lock_);
-  Apps::iterator it = apps_.find(app.first);
-  if (apps_.end() == it) {
-    LOG4CXX_INFO(logger_,
-                 "Adding application id "
-                     << app.first << " Application registered: " << app.second);
-    apps_.insert(app);
-    return;
+  bool is_need_handle_wrong_language = false;
+  {
+    sync_primitives::AutoLock lock(apps_lock_);
+    Apps::iterator it = apps_.find(app.first);
+    if (apps_.end() == it) {
+      LOG4CXX_INFO(logger_,
+                   "Adding application id "
+                       << app.first
+                       << " Application registered: " << app.second);
+      apps_.insert(app);
+      return;
+    }
```

Comment:
@dtrunov btw you don't need to do all this
```
Apps::iterator it = apps_.find(app.first);
if (apps_.end() == it) {
  LOG4CXX_INFO(logger_,
       "Adding application id "
       << app.first
       << " Application registered: " << app.second);
  apps_.insert(app);
  return;
}
```

because `insert()` will return false if key already present